---

title: bionlp


keywords: fastai
sidebar: home_sidebar

summary: "useful fxns for DL models with bio seqs"
description: "useful fxns for DL models with bio seqs"
nb_path: "00_core.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 00_core.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">bionlp_imports</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="gen_bio_vocab" class="doc_header"><code>gen_bio_vocab</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L14" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>gen_bio_vocab</code>(<strong><code>frameshift</code></strong>=<em><code>True</code></em>, <strong><code>kmer</code></strong>=<em><code>1</code></em>, <strong><code>seq</code></strong>=<em><code>'DNA'</code></em>)</p>
</blockquote>
<p>Makes a dictionary for mapping a raw sequence
to one hot encoding</p>
<p>seq = DNA or PROTEIN</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gen_bio_vocab</span><span class="p">(</span><span class="n">seq</span><span class="o">=</span><span class="s1">&#39;PROTEIN&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;A&#39;: 1,
 &#39;R&#39;: 2,
 &#39;N&#39;: 3,
 &#39;D&#39;: 4,
 &#39;B&#39;: 5,
 &#39;C&#39;: 6,
 &#39;E&#39;: 7,
 &#39;Q&#39;: 8,
 &#39;Z&#39;: 9,
 &#39;G&#39;: 10,
 &#39;H&#39;: 11,
 &#39;I&#39;: 12,
 &#39;L&#39;: 13,
 &#39;K&#39;: 14,
 &#39;M&#39;: 15,
 &#39;F&#39;: 16,
 &#39;P&#39;: 17,
 &#39;S&#39;: 18,
 &#39;T&#39;: 19,
 &#39;W&#39;: 20,
 &#39;Y&#39;: 21,
 &#39;V&#39;: 22}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="pred_vs_truth" class="doc_header"><code>pred_vs_truth</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L44" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>pred_vs_truth</code>(<strong><code>model</code></strong>, <strong><code>dataloader</code></strong>)</p>
</blockquote>
<p>takes a trained model and runs predictions over the
given dataloader.</p>
<p>plots hist of the ground truths vs predictions</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="raw2token" class="doc_header"><code>raw2token</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L110" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>raw2token</code>(<strong><code>seq</code></strong>)</p>
</blockquote>
<p>takes in raw sequence then converts it to tokenized version</p>
<p>'ATT' -&gt; [1 2 2]</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="token2hot" class="doc_header"><code>token2hot</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L119" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>token2hot</code>(<strong><code>seq</code></strong>, <strong><code>max_length</code></strong>, <strong><code>bio_dict</code></strong>)</p>
</blockquote>
<p>takes in tokenized sequences and returns 1-hot encoded</p>
<p>[1 2 2] -&gt; [1 0 0 0], [0 1 0 0 ], [0 1 0 0]</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="hot2token" class="doc_header"><code>hot2token</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L130" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>hot2token</code>(<strong><code>seq</code></strong>)</p>
</blockquote>
<p>one-hot to tokens</p>
<p>input:single sequence</p>
<p>e.g. [[0 0 1 0], [0 0 0 1] , [0 1 0 0]] -&gt; [2 3 1]</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="prep_seqs" class="doc_header"><code>prep_seqs</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L139" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>prep_seqs</code>(<strong><code>seq</code></strong>, <strong><code>max_length</code></strong>=<em><code>10</code></em>)</p>
</blockquote>
<p>takes in a raw seq 'ATTATA' -&gt; one hot encodes to specified max_len</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="DnaDataset" class="doc_header"><code>class</code> <code>DnaDataset</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L149" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>DnaDataset</code>(<strong>*<code>args</code></strong>, <strong>**<code>kwds</code></strong>) :: <code>Dataset</code></p>
</blockquote>
<p>takes in dataframe with ['Seqs'] and ['Exp']</p>
<p>returns pytorch dataset</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="StepByStep" class="doc_header"><code>class</code> <code>StepByStep</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L185" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>StepByStep</code>(<strong><code>model</code></strong>, <strong><code>loss_fn</code></strong>, <strong><code>optimizer</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_Seqs_Exp_Dataset" class="doc_header"><code>generate_Seqs_Exp_Dataset</code><a href="https://github.com/cjgo/bionlp/tree/master/bionlp/core.py#L406" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_Seqs_Exp_Dataset</code>(<strong><code>num_seq</code></strong>, <strong><code>low</code></strong>, <strong><code>high</code></strong>)</p>
</blockquote>
<p>generates a DNA dataset with Expression(or whatever else) values
nucleotides/expressions will be drawn from 2 distinct distributions</p>
<p>input : num_seq -&gt; int: number of sequences
        length_seq -&gt; tuple(min/max length of sequences)</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="n">dist</span> <span class="o">=</span> <span class="n">dist</span> <span class="o">/</span> <span class="n">dist</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
<span class="n">dist</span> <span class="o">=</span> <span class="n">dist</span> <span class="o">*</span> <span class="mi">3</span>
<span class="n">dist</span> <span class="o">=</span> <span class="n">dist</span> <span class="o">//</span> <span class="mi">1</span>
<span class="n">dist</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="n">dist</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>array([2., 2., 1., 1., 3., 2., 3., 2., 2., 1., 1., 1., 2., 2., 1., 2., 2.,
       2., 2., 2., 1., 3., 2., 2., 2., 2., 1., 1., 2., 1., 2., 1., 2., 3.,
       3., 2., 2., 2., 1., 1., 1., 2., 1., 1., 2., 2., 2., 2., 2., 2., 2.,
       2., 3., 3., 1., 2., 2., 2., 2., 4., 2., 1., 2., 1., 2., 2., 2., 2.,
       1., 2., 2., 1., 1., 2., 2., 2., 1., 1., 3., 1., 1., 2., 2., 2., 2.,
       3., 2., 2., 1., 1., 1., 1., 2., 2., 1., 1., 2., 2., 2., 2.])</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

